fem/fe.cpp

```diff
@@ -6566,61 +6591,285 @@ const double *Poly_1D::GetPoints(const int p, const int type)
+         delete it->second;
       }
-      delete it->second;
-   }
 
-   for (std::map<int, Array<Basis*>*>::iterator it = bases_container.begin();
-        it != bases_container.end() ; ++it )
-   {
-      Array<Basis*>& bases = *it->second;
-      for ( int i = 0 ; i < bases.Size() ; ++i )
+      for (std::map<int, Array<Basis*>*>::iterator it = basisContainer.begin();
+           it != basisContainer.end() ; ++it )
       {
-         delete bases[i];
+         Array<Basis*>& bases = *it->second;
+         for ( int i = 0 ; i < bases.Size() ; ++i )
+         {
+            delete bases[i];
+         }
+         delete it->second;
       }
-      delete it->second;
    }
 }
 
 Poly_1D poly1d;
 Array2D<int> Poly_1D::binom;
 
+TensorBasisElement::TensorBasisElement(const int dims,
+                                       const int p,
+                                       const int dofs,
+                                       const int type,
+                                       const int btype) :
+   pt_type(type),
+   dof_map(dofs),
+   basis1d(poly1d.GetBasis(p, pt_type, btype))
+{
+
+   if (FiniteElement::IsClosedType(pt_type))
+   {
+      switch (dims)
+      {
+         case 1:
+         {
+            dof_map[0] = 0;
+            dof_map[p] = 1;
+            for (int i = 1; i < p; i++)
+            {
+               dof_map[i] = i+1;
+            }
+            break;
+         }
+         case 2:
+         {
+            const int p1 = p + 1;
+
+            // vertices
+            dof_map[0 + 0*p1] = 0;
+            dof_map[p + 0*p1] = 1;
+            dof_map[p + p*p1] = 2;
+            dof_map[0 + p*p1] = 3;
+
+            // edges
+            int o = 4;
+            for (int i = 1; i < p; i++)
+            {
+               dof_map[i + 0*p1] = o++;
+            }
+            for (int i = 1; i < p; i++)
+            {
+               dof_map[p + i*p1] = o++;
+            }
+            for (int i = 1; i < p; i++)
+            {
+               dof_map[(p-i) + p*p1] = o++;
+            }
+            for (int i = 1; i < p; i++)
+            {
+               dof_map[0 + (p-i)*p1] = o++;
+            }
+
+            // interior
+            for (int j = 1; j < p; j++)
+            {
+               for (int i = 1; i < p; i++)
+               {
+                  dof_map[i + j*p1] = o++;
+               }
+            }
+            break;
+         }
+         case 3:
+         {
+            const int p1 = p + 1;
+
+            // vertices
+            dof_map[0 + (0 + 0*p1)*p1] = 0;
+            dof_map[p + (0 + 0*p1)*p1] = 1;
+            dof_map[p + (p + 0*p1)*p1] = 2;
+            dof_map[0 + (p + 0*p1)*p1] = 3;
+            dof_map[0 + (0 + p*p1)*p1] = 4;
+            dof_map[p + (0 + p*p1)*p1] = 5;
+            dof_map[p + (p + p*p1)*p1] = 6;
+            dof_map[0 + (p + p*p1)*p1] = 7;
+
+            // edges (see Hexahedron::edges in mesh/hexahedron.cpp)
+            int o = 8;
+            for (int i = 1; i < p; i++)
```

Comment:
Shouldn't we replace this with a macro or an inline function? If there is ever a bug in this, we're unlikely to find it. Macros are not safe in C++ though.